### DISCUSSION TITLE
android PAG 缓存目录可否移动到内部私有目录下，部分手机无法获取外部私有目录

### DISCUSSION THREAD

**Discussion by zengke328** [Tencent/libpag#2252] [Q&A] [answered]
上次反馈的问题 https://github.com/Tencent/libpag/discussions/2242。官方demo可复现PAGImageView无法加载assets 动画。
经排查，系部分三星手机无法获取到应用私有的外部存储目录，如context.getExternalCacheDir()、context.getExternalFilesDir()。

看源码PAGDiskCache类中返回的缓存目录会优先使用context.getExternalCacheDir()。能否只使用context.getCacheDir()目录来解决该问题。
![image](https://github.com/Tencent/libpag/assets/22447723/c69d4826-41da-43f2-98fd-9f4bb727c4d3)

**Comment by kevingpqi123** [comment]
能否提供下复现机型上 context.getExternalCacheDir() 上返回的是什么内容，如果返回的是空，会切换到 context.getCacheDir()

**Reply by zengke328** [reply]
context.getExternalCacheDir()可以获取到内容。抱歉，是我昨天描述的不够准确，我们排查到的是部分三星手机没有写入私有外部存储目录的权限。如果我们的Pag缓存也是存储在外部存储目录，也会存在这个问题。
ps：刚我在三星云真机调试，使用Galaxy Z Flip5没能复现（云真机看不到手机系统信息），推测和系统有关。

**Reply by kevingpqi123** [reply]
那建议你们提供下具体的可以复现的机型，我们只有定位到最终的原因，才好在这里添加兼容逻辑

**Reply by Yetland** [reply]
云真机s23可以复现

**Reply by Yetland** [reply]
我在云真机上测试了一下，和cache位置没有关系。即使使用内部缓存，也无法加载。复现的机型：https://developer.samsung.com/remotetestlab/devices/125/galaxy-s 选择s23-fe。

**Reply by kevingpqi123** [reply, marked as answer]
针对三星部分手机 PAGImageView 渲染空白的问题，我们分析了下原因，结论如下：
（1）该问题和缓存路径没有太大关系；
（2）出现问题的原因是三星部分机型 HardwareBuffer 相关实现可能存在一些问题，对于 PAGImageView 而言，pag 在播放动画的过程中，创建了一个 Bitmap，该 Bitmap 通过 Bitmap.wrapHardwareBuffer 的方式创建，Bitmap 占用的内存来源于创建的 HardwareBuffer，播放的过程中会不断将渲染的数据 copy 至创建的HardwareBuffer，然后将 Bitmap 渲染出来，但是在三星的部分机型上 Bitmap 的宽高可以正常读取，却无法正常显示，目前已拉群将该问题反馈至三星

对于 pag 而言，如果这里做处理就要加机型判断，我们倾向于等待三星的问题修复，问题修复后会第一时间知会大家
